lib/Transforms/AffineDataCopyGeneration.cpp

```diff
@@ -392,50 +453,75 @@ bool DmaGeneration::generateDma(const MemRefRegion &region, Block *block,
   }
 
   // Record the last operation just before the point where we insert the
-  // outgoing DMAs. We later do the memref replacement later only in [begin,
-  // postDomFilter] so that the original memref's in the DMA ops themselves
-  // don't get replaced.
+  // copy out's. We later do the memref replacement later only in [begin,
+  // postDomFilter] so that the original memref's in the data movement code
+  // themselves don't get replaced.
   auto postDomFilter = std::prev(end);
 
   // Create fully composed affine maps for each memref.
   auto memAffineMap = b.getMultiDimIdentityMap(memIndices.size());
   fullyComposeAffineMapAndOperands(&memAffineMap, &memIndices);
   auto bufAffineMap = b.getMultiDimIdentityMap(bufIndices.size());
   fullyComposeAffineMapAndOperands(&bufAffineMap, &bufIndices);
-  SmallVector<Value *, 4> tagIndices({zeroIndex});
-  auto tagAffineMap = b.getMultiDimIdentityMap(tagIndices.size());
-  fullyComposeAffineMapAndOperands(&tagAffineMap, &tagIndices);
-  if (!region.isWrite()) {
-    // DMA non-blocking read from original buffer to fast buffer.
-    b.create<AffineDmaStartOp>(loc, memref, memAffineMap, memIndices,
-                               fastMemRef, bufAffineMap, bufIndices, tagMemRef,
-                               tagAffineMap, tagIndices, numElementsSSA, stride,
-                               numEltPerStride);
+
+  if (!clDma) {
+    auto copyNest = generatePointWiseCopy(loc, memref, fastMemRef, memIndices,
```

Comment:
Not in this PR, but this makes me wonder if in the future we could separate this logic out. For example, we could create a higher level AffineDataTransferOp(%src_memref, %dst_memref, %num_elements) which could be lowered at a later time to either DMAs or copy loop nests.

I'm wondering if this might help us sort out two concerns where are tangled together in this pass:
1) Computing the union of memref regions and loop nest depth at which to insert copies.
2) Transforming the IR to implement these copies.

Alternatively, we could split this PR into 3 things:
1) DataTransferAnalysis pass (which returns memref region unions and loop depths)
2) DmaGenerationPass
3) LoopNestCopyGenerationPass

Reply:
Yes, before I started I implementing this, I was actually thinking about having such a transfer op that could be expanded out either to point-wise copies or DMAs - the op actually will carry all the information to do this expansion. But in terms of code reuse and separation, the current code is already pretty close to that (although all in the same file). Also, I didn't right now see the need for delaying that expansion, but if we do, adding such an op appears to be the way to go.